### ISSUE TITLE
javax.naming.SizeLimitExceededException: [LDAP: error code 4 - Sizelimit Exceeded];

### ISSUE THREAD

**Issue by inits** [gocd/gocd-ldap-authentication-plugin#38]
``` 
private LdapConfiguration ldapConfiguration;
 private final int MAX_AUTHENTICATION_RESULT = 1;
```

why set sizelimit to 1 ?

**Comment by ketan** [comment]
Ideally — when you're looking up a user in an LDAP tree, you should find a unique user matching the criteria specified. Otherwise, it's ambiguous which user is attempting to login.

Can you explain why you think the sizelimit needs to be something else?

**Comment by inits** [comment]
sorry for that, that my mistake, i think i found a bug , i have two username ,one is puqiaoming, this username can login, another is qiaoming ,when i login with qiaoming have error below:

```
2018-04-05 14:31:24,642  INFO [qtp1301664418-27] p.c.g.a.l.c.g.a.l.LdapPlugin:72 [plugin-cd.go.authentication.ldap] - [Authenticate] Authenticating User: qiaoming using auth_config: lwork
2018-04-05 14:31:24,657 ERROR [qtp1301664418-27] p.c.g.a.l.c.g.a.l.LdapPlugin:127 [plugin-cd.go.authentication.ldap] - [LDAP: error code 4 - Sizelimit Exceeded]
javax.naming.SizeLimitExceededException: [LDAP: error code 4 - Sizelimit Exceeded]
	at com.sun.jndi.ldap.LdapCtx.mapErrorCode(LdapCtx.java:3206)
	at com.sun.jndi.ldap.LdapCtx.processReturnCode(LdapCtx.java:3100)
	at com.sun.jndi.ldap.LdapCtx.processReturnCode(LdapCtx.java:2891)
	at com.sun.jndi.ldap.AbstractLdapNamingEnumeration.getNextBatch(AbstractLdapNamingEnumeration.java:148)
	at com.sun.jndi.ldap.AbstractLdapNamingEnumeration.hasMoreImpl(AbstractLdapNamingEnumeration.java:217)
	at com.sun.jndi.ldap.AbstractLdapNamingEnumeration.hasMore(AbstractLdapNamingEnumeration.java:189)
	at cd.go.framework.ldap.Ldap.search(Ldap.java:105)
	at cd.go.framework.ldap.Ldap.authenticate(Ldap.java:46)
	at cd.go.authentication.ldap.LdapAuthenticator.authenticateWithAuthConfig(LdapAuthenticator.java:58)
	at cd.go.authentication.ldap.LdapAuthenticator.authenticate(LdapAuthenticator.java:44)
	at cd.go.authentication.ldap.executor.UserAuthenticationExecutor.execute(UserAuthenticationExecutor.java:50)
	at cd.go.authentication.ldap.LdapPlugin.handle(LdapPlugin.java:69)
	at com.thoughtworks.go.plugin.infra.DefaultPluginManager$2.execute(DefaultPluginManager.java:167)
	at com.thoughtworks.go.plugin.infra.DefaultPluginManager$2.execute(DefaultPluginManager.java:162)
	at com.thoughtworks.go.plugin.infra.FelixGoPluginOSGiFramework.executeActionOnTheService(FelixGoPluginOSGiFramework.java:316)
	at com.thoughtworks.go.plugin.infra.FelixGoPluginOSGiFramework.doOn(FelixGoPluginOSGiFramework.java:246)
	at com.thoughtworks.go.plugin.infra.DefaultPluginManager.submitTo(DefaultPluginManager.java:162)
	at com.thoughtworks.go.plugin.access.PluginRequestHelper.submitRequest(PluginRequestHelper.java:48)
	at com.thoughtworks.go.plugin.access.authorization.AuthorizationExtension.authenticateUser(AuthorizationExtension.java:122)
	at com.thoughtworks.go.server.security.providers.PluginAuthenticationProvider.getUserDetailsFromAuthorizationPlugins(PluginAuthenticationProvider.java:119)

```

**Comment by ketan** [comment]
Thanks for the bug report. @maheshp @bdpiparva — can one of you look into this?

**Comment by bdpiparva** [comment]
@inits --- Please can you provide what value you have used for user `UserLoginFilter` in your auth config.

**Comment by inits** [comment]
is default below 
(|(sAMAccountName=*{0}*)(uid=*{0}*)(cn=*{0}*)(mail=*{0}*)(otherMailbox=*{0}*))

**Comment by ketan** [comment]
@bdpiparva and @maheshp — should we use `*` around the filters `(|(sAMAccountName=*{0}*)(uid=*{0}*)(cn=*{0}*)(mail=*{0}*)(otherMailbox=*{0}*))`. I'd think it should simply be `(|(sAMAccountName={0})(uid={0})(cn={0})(mail={0})(otherMailbox={0}))`

**Comment by bdpiparva** [comment]
@ketan --- No we don't need * around filters.

@inits  -- Please can you change it to `(|(sAMAccountName={0})(uid={0})(cn={0})(mail={0})(otherMailbox={0}))`. It should resolve your issue.

**Comment by bdpiparva** [comment]
@ketan --- https://github.com/gocd/gocd-ldap-authentication-plugin/issues/38#issuecomment-380822424 there is no default value for `UserLoginFilter`. Ignore what I said earlier.

**Comment by bdpiparva** [comment]
Yes, that is what I am planning. 
1. Tooltip on the configuration page 
2. Detect if filter contains any `*{0}*`, `*{0}`, or `{0}*` then show error.

**Comment by arvindsv** [comment]
The documentation definitely needs to change.

**Comment by arvindsv** [comment]
@bdpiparva and I talked. Here are some notes:

1. In `UserSearchFilter` `*{0}*` is appropriate. This is so that "Add user" functionality and others such as that will work, and find users by partial name.

2. In `UserLoginFilter` `*{0}*` is **not** appropriate. It has the potential of finding mutiple users for a given username, when logging in. This is what is happening in the original exception.

3. None of our documentation needs to change, because we do have it correct. We will consider a warning in the tooltip message of `UserLoginFilter` warning against using `*{0}*`. Something like: ```It is not recommended to have *{0}* in this field as it can match other users.```

4. We were able to reproduce this in a test and we will be showing a better exception message. In essence, we will be differentiating between a hard limit of 1 match during login and a soft limit of 10 or 15 matches for "Add user" functionality, which should take care of this problem. In one LDAP search base, if a user, while logging in, matches multiple LDAP records, then an error will be logged.
